packages/devtools_app/lib/src/charts/flutter/flame_chart.dart

```diff
@@ -14,6 +18,179 @@ const double sectionSpacing = 15.0;
 const double topOffset = rowHeightWithPadding;
 const double sideInset = 70.0;
 
+// TODO(kenz): consider cleaning up by changing to a flame chart code to use a
+// composition pattern instead of a class extension pattern.
+abstract class FlameChart<T, V extends FlameChartNodeDataMixin>
+    extends StatefulWidget {
+  const FlameChart(
+    this.data, {
+    @required this.duration,
+    @required this.height,
+    @required this.totalStartingWidth,
+    @required this.startInset,
+    @required this.selectionNotifier,
+    @required this.onSelection,
+  });
+
+  final T data;
+
+  final Duration duration;
+
+  final double totalStartingWidth;
+
+  final double height;
+
+  final double startInset;
+
+  final ValueListenable<V> selectionNotifier;
+
+  final void Function(V event) onSelection;
+
+  double get startingContentWidth =>
+      totalStartingWidth - startInset - sideInset;
+}
+
+abstract class FlameChartState<T extends FlameChart> extends State<T>
+    with AutoDisposeMixin, FlameChartColorMixin {
+  // The "top" positional value for each flame chart node will be 0.0 because
+  // each node is positioned inside its own stack.
+  final flameChartNodeTop = 0.0;
+
+  LinkedScrollControllerGroup _linkedScrollControllerGroup;
+
+  List<FlameChartRow> rows = [];
+
+  /// Starting pixels per microsecond in order to fit all the data in view at
+  /// start.
+  double get startingPxPerMicro =>
+      widget.startingContentWidth / widget.data.time.duration.inMicroseconds;
+
+  int get startTimeOffset => widget.data.time.start.inMicroseconds;
+
+  @override
+  void initState() {
+    super.initState();
+    initFlameChartElements();
+    addAutoDisposeListener(widget.selectionNotifier);
+    _linkedScrollControllerGroup = LinkedScrollControllerGroup();
+  }
+
+  @override
+  void didUpdateWidget(T oldWidget) {
+    if (widget.data != oldWidget.data) {
+      _linkedScrollControllerGroup.resetScroll();
+      initFlameChartElements();
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return ListView.builder(
+          itemCount: rows.length,
+          itemBuilder: (context, index) {
+            return ScrollingFlameChartRow(
+              linkedScrollControllerGroup: _linkedScrollControllerGroup,
+              sortedNodes: rows[index].nodes
+                ..sort((a, b) => a.rect.left.compareTo(b.rect.left)),
```

Comment:
A list sort during build is going to make rebuilds very expensive.

Is it possible to precompute this?